sample-prep/Biomek_pooling_VI.py

```diff
@@ -29,107 +43,287 @@ def main(process_id, file_id):
+                sample_volumes[i] = max_vol_available
+                volume_threshold_plotcolor[i] = "yellow"
 
-    #write csv-file
+    if enforce_min_volume == True:
+        if min_volume_in_list < min_vol_required:
+            volume_threshold_plotcolor[min_volume_in_list_index] = "yellow"
+
+    # DEFINE BIOMEK DECK PLATE POSITIONS (LIBLABWARE)
+    # create list with biomek LibLabware variable for sorting plate position on biomek deck 
+    # (plate 1 = position "Libraries 1", plate 2 = position "Libraries 2")
+    # no duplicate values allowed in set, so can extract unique container names by redefining to set. 
+
+    input_container_name_unique = sorted(list(set(input_container_name)))
+    print(input_container_name_unique)
+    libLabware = []
+    for i in range(len(input_container_name)):
+        if input_container_name[i] == input_container_name_unique[0]:
+            libLabware.append("Libraries1")
+        elif input_container_name[i] == input_container_name_unique[1]:
+            libLabware.append("Libraries2")
+        else:
+            print("Error assigning Biomek LibLabware positions. Contact system admin")
+            sys.exit(1)
+
+    #WRITE CSV FILE 
+    # using data loaded above. Looping over all inputs and writing line-by-line to the csv-file
+
+    total_sample_number = len(sample_name)
 
     with open(file_id + "-Biomek_pooling.csv", "w") as file:
-        file.write("LibLabware,LibWellPos,DestLabware,DestWellPos,LibVol,SampleName,Molarity,Species,GenomeSize,ProjectAccount" + '\n')
+        file.write("LibLabware,LibWellPos,DestLabware,DestWellPos,LibVol,SampleName,Molarity,Species,GenomeSize,PlateID,fmol/Mbp-actual" + '\n')
         for i in range(total_sample_number):
             file.write(
-                "Libraries" + ',' + 
+                str(libLabware[i]) + ',' + 
                 str(re.sub(r':','',source_location[i][1])) + ',' + 
                 'LibraryPool'  + ',' + 
                 "D1" + "," + 
-                str(sample_volumes[i]) + "," + 
+                str(round(sample_volumes[i],1)) + "," + 
                 str(sample_name[i]) + "," + 
                 str(molarity[i]) + "," +
                 str(species[i]) + "," + 
                 str(genome_size[i]) + "," +
-                str(project_account[i]) + "," + '\n')
+                str(input_container_name[i]) + "," +
+                str(round((sample_volumes[i]*molarity[i])/genome_size[i],0))+ '\n')
     
-    #set process UDFs - Total, Max and Min volumes in pooling list
+    #SET PROCESS UDFs - Total, Max and Min volumes.
+    # Will display on the step after running the script. Makes it easy to quickly evaluate if you need to adjust fmol/Mbp and rerun
 
     process.udf["Total Pool Volume"] = round(sum(sample_volumes),2)
     process.udf["Max Pooling Volume"] = round(max(sample_volumes),2)
     process.udf["Min Pooling Volume"] = round(min(sample_volumes),2)
     process.put()
 
```

Comment:
with the put here, do you put stuff back into the db? If so, is it the above udfs you are putting things back into?

Reply:
And do they exist already, or do you create them by puting them?

Reply:
Yes, I put back all the process values I have cached into my script. Isnt a 'put' by definition updating an existing entity? 'Post' would be creating a new one. But yes, they do exist. If not you would get an error message. And I do not think you can create/post new udfs (not that I can imagine that you would ever want to do that through api)